[RFC] JHipster Kubernetes Operator #10053

salaboy opened the pull request:
[skip ci]

This RFC related to https://github.com/jhipster/generator-jhipster/issues/10053

I will appreciate feedback and help to refine the RFC with more details. I am still applying feedback on the following repository to make the Operator (first iteration) simpler.

salaboy commented:
@saturnism @jdubois @deepu105 @pascalgrimaud @PierreBesson  I would appreciate feedback on this so I can move the Operator forward. I am doing some more work here: https://github.com/salaboy/jhipster-operator
As I am a firm believer of having something working will help us to richer conversations. But I am happy to move that work to a more formal repo for everyone to keep track of it.

PierreBesson commented:
Great work! I agree let's move your project to the jhipster or hipster-labs github org and create a build pipeline to build and publish your jhipster operator to docker hub.

@salaboy the video has just been published : https://youtu.be/9iqTtwptTT8?list=PL6IFaLdAcgE3aSgmRyhi6eULdlrwjA5sm 
Can you add the link to the rfc.

salaboy commented:
@PierreBesson can you please create the repo so I can start moving the code? 
I will link the video to my blog post and the RFC :)

deepu105 commented:
While this is really interesting and super cool as a concept, we need a wider discussion before you actually spend more time on it. I'm gonna wear my pragmatic hat now and ask some questions we need to answer first
1. This definitely is cool but is this really necessary for a JHipster application.
2. Are we rebuilding any features provided by tools like Istio(coz many of the proposed features looks very similar to what it does)
3. While a JHipster CRD is cool it also means more tie in into JHipster in terms of deployment architecture is that necessary?

So lets first discuss with the community

jdubois commented:
This definitely looks very cool, but I'm not the most competent on this part -> do you want me to send a few tweets, maybe a Twitter poll, so gather more community feedback?

deepu105 commented:
Yup more eyes wouldnt hurt

Thanks & Regards,
Deepu


On Mon, Jul 8, 2019 at 5:26 PM Julien Dubois <notifications@github.com>
wrote:

> This definitely looks very cool, but I'm not the most competent on this
> part -> do you want me to send a few tweets, maybe a Twitter poll, so
> gather more community feedback?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jhipster/generator-jhipster/pull/10054?email_source=notifications&email_token=AAIOKFZNPSQSHOYOKAXBWILP6NMDNA5CNFSM4H6ZSPSKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZNOFTA#issuecomment-509272780>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIOKFY5XHP7CA2JAREZJA3P6NMDNANCNFSM4H6ZSPSA>
> .
>

jdubois commented:
Sorry, the RFC process is new to me -> isn't discussing the whole point of the RFC? Then, shouldn't this be merged first?

deepu105 commented:
Ya i'm not sure as well, do we merge first and then discuss or discuss before merging?

For me makes sense to discuss first and merge only when approved

PierreBesson commented:
Let's discuss first and merge once we agree on a first version.

salaboy commented:
@deepu105 I will add more details, so it is clear. 
In general I've seen this feeling of .. "this new "thing" is going to change everything".. while in fact it is quite the opposite (it is just an add on, if you are running in k8s to make your life easier), and it can be built in a way where it is optional and deployed later on if you don't want to start with it. 
Until now,  the only way that I found to remove the fear of the unknown is to have something up and running where people can play with it and see the value. There is no need for commitment, but I would appreciate help on non-blocking it, just for the sake of the experiments. 
I will ping you guys when I add more details.. probably today during the day.

salaboy commented:
@jdubois I am happy to have more eyes on this, I will do my best to explain clearly in the RFC text what this is about and how it will work so we all can see the benefits

salaboy commented:
@deepu105 I've answered your questions in a FAQ section of the RFC.

I will change the diagrams, because I think that are causing confusion about changing the architecture.  The Operator doesn't change the architecture at all, it just extend it, and it should be an optional extension.

deepu105 commented:
Yes please. More details the better. I'm definitely not trying to block it
my best interests are for the community so if this adds value to the
community I'll be happy. And I'm playing the devil's advocate here.

On Tue, 9 Jul 2019, 10:39 am salaboy, <notifications@github.com> wrote:

> @deepu105 <https://github.com/deepu105> I've answered your questions in a
> FAQ section of the RFC.
>
> I will change the diagrams, because I think that are causing confusion
> about changing the architecture. The Operator doesn't change the
> architecture at all, it just extend it, and it should be an optional
> extension.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jhipster/generator-jhipster/pull/10054?email_source=notifications&email_token=AAIOKF5DZFY3IOESP2J2QKDP6RFENA5CNFSM4H6ZSPSKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZPRR2Y#issuecomment-509548779>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIOKF6R2BVHKB4NMGNQZSLP6RFENANCNFSM4H6ZSPSA>
> .
>

salaboy commented:
More details added @deepu105 @jdubois @PierreBesson @saturnism , I've added some images as well hoping to clarify the whole approach. I've also added a section stating what the Operator is **NOT** going to do. 
I like the idea of more questions for the FAQ section, because it definitely help to explain the scope of the work.

PierreBesson commented:
For me this is a very good idea. It clicked for me during @salaboy talk that this is the missing link to achieve a cohesive microservice stack on Kubernetes like we had on Spring Cloud. For example I have always wondered how should we setup the gateway/registries to dynamically find available microservices in Kubernetes like we had using Eureka/Consul discovery.

Indeed it seems that developping a custom operator is the only way to start to be kubernetes native and adapt our successful existing stack (registry/gateway/etc.) to implement microservices patterns on top of kubernetes.

PierreBesson commented on rfcs/1-jhipster-rfc-k8s-operator.md:
```diff
@@ -0,0 +1,161 @@
+  - Understand how things are wired together
+  - Understand and validate the Applications State as a whole
+- Provide advanced lifecycle management
+  - Shared infrastructure wiring
```
In this section I would add `Manage centralized configuration`, I would love to have an higher level abstraction to manage configuration in kubernetes. Currently we have the ConfigMap but it's not great for many reasons (for example, it does not trigger an update of the apps that depend on it).

salaboy replied:
Totally.. that is part of lifecycle management and understanding how things are wired together. 
I will add it there.

salaboy replied:
done

PierreBesson commented:
As it's been 2 weeks without additional discussion. We have no objection to this proposal. Let's try and see where it will take us. I don't have the rights to create a repo in the jhipster organisation, so I can't do it myself.

@pascalgrimaud can you create a new `jhipster-operator` repo and a new `operator-hipsters` group with write access to it that will include @salaboy.

salaboy commented:
@PierreBesson @pascalgrimaud can you guys please share the links when you have those? So I can put the code that I have in there..

pascalgrimaud commented:
It should be ok:
- https://github.com/jhipster/jhipster-operator

@salaboy : don't hesitate to tell me if you can't push to the project, you should have correct access
